### PR TITLE
DFAST-238 : Backward compatibility mode: Use apply_q[i] to flag discharges to be processed instead of discharge[i] is not None

### PR DESCRIPTION
**Pull request by hrajagers**
The "discharge[i] is None" check does not work for i = 0 since discharge[0] will never be equal to None, whereas the user may not have to specify any data for that discharge level, i.e. apply_q[0] may be True or False. Only if apply_q[0] is True this level needs to be processed.

### CODE REVIEW

File: dfastmi/batch/core.py [hunk 1]
```diff
@@ -483,21 +483,22 @@ def _get_root_dir(rootdir: Path) -> Path:
     return rootdir
 
 
-def count_discharges(discharges: Vector) -> int:
+def count_discharges(apply_q: Tuple[bool, bool, bool]) -> int:
```
Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:486]: Please use BoolVector

Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:486]: If this is a legacy method move to legacy.py. Do not do this in this issue (not in scope) but create a new issue to cleanup this file for improved maintenance.

File: dfastmi/batch/core.py [hunk 2]
```diff
@@ -483,21 +483,22 @@ def _get_root_dir(rootdir: Path) -> Path:
...
     """
-    Count the number of non-empty discharges.
+    Count the number of discharges to be used (apply_q = True).
 
     Arguments
     ---------
-    discharges : Vector
-        Tuple of (at most) three characteristic discharges (Q).
+    apply_q : Tuple[bool, bool, bool]
```
Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:492]: update the commented reflecting the new type

File: dfastmi/batch/core.py [hunk 3]
```diff
@@ -707,6 +708,7 @@ def write_report(
     tstag: float,
     q_fit: Tuple[float, float],
     discharges: Vector,
+    apply_q: Tuple[bool, bool, bool],
```
Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:711]: Use BoolVector type

File: dfastmi/batch/core.py [hunk 4]
```diff
@@ -483,21 +483,22 @@ def _get_root_dir(rootdir: Path) -> Path:
...
     Returns
     -------
     n : int
         Number of non-empty discharges.
     """
-    return sum([q is not None for q in discharges])
+    return sum(apply_q)
 
 
 def get_filenames(
```
Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:504]: If this is a legacy method move to legacy.py. Do not do this in this issue (not in scope) but create a new issue to cleanup this file for improved maintenance.

File: dfastmi/cli.py [hunk 5]
```diff
@@ -573,6 +574,7 @@ def _write_report_nodata(
     tstag: float,
     q_fit: Tuple[float, float],
     discharges: QRuns,
+    apply_q: Tuple[bool, bool, bool],
```
Comment (ralph-peelen-deltares) [inline dfastmi/cli.py:577]: Use BoolVector Type

File: dfastmi/batch/core.py [hunk 6]
```diff
@@ -707,6 +708,7 @@ def write_report(
     tstag: float,
```
Comment (ralph-peelen-deltares) [inline dfastmi/batch/core.py:708]: If this is a legacy method move to legacy.py. Do not do this in this issue (not in scope) but create a new issue to cleanup this file for improved maintenance.

**Review by ralph-peelen-deltares — CHANGES_REQUESTED** [review]
Please note unit test are failing, but seems to be a teamcity issue. Please run again on different (earlier successful) agent

File: dfastmi/batch/core.py (hunk 1 above)
Comment (hrajagers) [reply dfastmi/batch/core.py]: Done. However, `apply_q` is a `Tuple` and `BoolVector` is a shorthand for a `List` type ... I see that this is mixed up at other locations as well.
I've made the use of `BoolVector` now consistent, but created a new issue DFAST-236 to address the type inconsistency. I needed tuples in the past for stricter type checking by the `pypy` static analyzer, but lists should also work.

File: dfastmi/batch/core.py (hunk 2 above)
Comment (hrajagers) [reply dfastmi/batch/core.py]: done

File: dfastmi/batch/core.py (hunk 3 above)
Comment (hrajagers) [reply dfastmi/batch/core.py]: done

File: dfastmi/cli.py (hunk 5 above)
Comment (hrajagers) [reply dfastmi/cli.py]: done

File: dfastmi/batch/core.py (hunk 6 above)
Comment (hrajagers) [reply dfastmi/batch/core.py:708]: New issue created DFAST-237.

File: dfastmi/batch/core.py (hunk 4 above)
Comment (hrajagers) [reply dfastmi/batch/core.py:504]: Included in issue DFAST-237.

File: dfastmi/batch/core.py (hunk 1 above)
Comment (hrajagers) [reply dfastmi/batch/core.py]: Included in issue DFAST-237.

File: docs/chapters/backward_compatibility.tex [hunk 7]
```diff
@@ -1279,7 +1279,7 @@ \subsection{Analysis configuration file}\label{app-v1:config}
 \keyw{General} & \keyw{Mode} & \keyw{WAQUA export} or \keyw{D-Flow FM map} (the latter is the default) \\
 \keyw{General} & \keyw{Branch} & Name of the selected branch \\
 \keyw{General} & \keyw{Reach} & Name of the selected reach \\
-\keyw{General} & \keyw{QMin} & Minimum discharge \unitbrackets{m\textsuperscript{3}/s} at which intervention becomes active \\
```
Comment (ralph-peelen-deltares) [inline docs/chapters/backward_compatibility.tex:1282]: QMin is dus nooit een key geweest? Was altijd al QThreshold?

Comment (ralph-peelen-deltares) [reply docs/chapters/backward_compatibility.tex:1282]: besproken, klopt

File: dfastmi/cli.py [hunk 8]
```diff
@@ -573,6 +574,7 @@ def _write_report_nodata(
     tstag: float,
     q_fit: Tuple[float, float],
     discharges: QRuns,
+    apply_q: BoolVector,
```
[the remark this answers is not in the document]
Comment (ralph-peelen-deltares) [reply dfastmi/cli.py:577]: en dan hier een object meegeven:
''' 
discharge :  Discharges[]
'''
Is Qruns en BoolVector dan nog nodig?

nogmaals is voor DFAST-237

**Review by ralph-peelen-deltares — APPROVED** [review]
please see comments... seems good but look at using (custom) objects instead of using custom types (related to each other)

File: dfastmi/cli.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hrajagers) [reply dfastmi/cli.py:577]: That may indeed be an option. I've added it to DFAST-237 with the comment that this is legacy code, so we need to evaluate whether it's still worth the effort.

**Pull request merged**